tools/clang/lib/SPIRV/DeclResultIdMapper.cpp

```diff
@@ -1963,6 +1963,47 @@ bool DeclResultIdMapper::createStageVars(
     // function/parameter/variable. All are DeclaratorDecls.
     stageVarInstructions[cast<DeclaratorDecl>(decl)] = varInstr;
 
+    // Special case: The DX12 SV_InstanceID always counts from 0, even if the
+    // StartInstanceLocation parameter is non-zero. gl_InstanceIndex, however,
+    // starts from firstInstance. Thus it doesn't emulate actual DX12 shader
+    // behavior. To make it equivalent, SPIR-V codegen should emit:
+    // SV_InstanceID = gl_InstanceIndex - gl_BaseInstance
+    // Unfortunately, this means that there is no 1-to-1 mapping of the HLSL
+    // semantic to the SPIR-V builtin. As a result, we have to manually create
+    // a second stage variable for this specific case.
+    if (asInput && semanticKind == hlsl::Semantic::Kind::InstanceID &&
+        sigPointKind == hlsl::SigPoint::Kind::VSIn) {
+      // The above call to createSpirvStageVar creates the gl_InstanceIndex.
+      // We should now manually create the gl_BaseInstance variable and do the
+      // subtraction.
+      auto *instanceIndexVar = varInstr;
+      auto *baseInstanceVar = spvBuilder.addStageBuiltinVar(
+          type, spv::StorageClass::Input, spv::BuiltIn::BaseInstance,
+          decl->hasAttr<HLSLPreciseAttr>(), semanticToUse->loc);
+      StageVar stageVar2(sigPoint, *semanticToUse, builtinAttr, evalType,
+                         getLocationCount(astContext, type));
+      stageVar2.setSpirvInstr(baseInstanceVar);
+      stageVar2.setLocationAttr(decl->getAttr<VKLocationAttr>());
+      stageVar2.setIndexAttr(decl->getAttr<VKIndexAttr>());
+      stageVar2.setIsSpirvBuiltin();
+      stageVars.push_back(stageVar2);
+
+      // Do the math.
```

Comment:
Better to explicitly explain `SPIR-V code fore 'SV_InstanceID = gl_InstanceIndex - gl_BaseInstance'` instead of saying "Do the math"?
For me, it will save the time to read the code but I am not sure how other people feel.